### PR TITLE
Add admission control to support default service plans. Resolves issue 1111

### LINKED ISSUE #1111: Make "plan" optional when there is only one plan to choose from
Per the discussion in #1010 and our design meeting on Aug 7th, we agreed that:
- ServiceInstance.Plan should be made optional, and when "plan" is not specified then:
- If there is only one plan for the specified ServiceClass then we'll default to that plan
- If there is more than one plan for the specified ServiceClass then we'll reject the request to create the new ServiceInstance 

Related to this we should:
- change the docs for UPS in the walkthru to remove the use of "plan" just to make a nicer U

Closes #1010

### PR DESCRIPTION
**Pull request by vaikas-google**
Closes #1111

### CODE REVIEW

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	if len(instance.Spec.PlanName) > 0 {
+		return nil
+	}
+
+	sc, err := d.scLister.Get(instance.Spec.ServiceClassName)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return errors.NewInternalError(err)
+		}
+		msg := fmt.Sprintf("ServiceClass %q does not exist, PlanName must be specified", instance.Spec.ServiceClassName)
```
Comment (kibbles-n-bytes) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go]: The "PlanName must be specified" part of the error seems irrelevant to this particular situation where we simply could not get the service class.

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission_test.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+	scList.Items = append(scList.Items, *sc)
+
+	fakeClient.AddReactor("list", "serviceclasses", func(action core.Action) (bool, runtime.Object, error) {
+		return true, scList, nil
+	})
+	return fakeClient
+}
+
+// newInstance returns a new instance for the specified namespace.
+func newInstance(namespace string) servicecatalog.Instance {
```
Comment (kibbles-n-bytes) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission_test.go]: This function is duplicated from the namespace lifecycle admission controller tests. It seems to me it'd be useful to have a `util` package shared across all admission controller tests, as functions such as these seem to be common, and we will be adding more admission controllers as time goes on.

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
+/*
```
Comment (kibbles-n-bytes) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go]: This admission controller's behavior as it stands would be completely undocumented. Could we get a short message somewhere in our docs that mentions it?

I guess the most appropriate place would be: https://github.com/kubernetes-incubator/service-catalog/blob/master/docs/design.md#creating-a-service-instance . It's rough around the edges for now, but at least having a mention would remind us to include it in a rewrite should we do so later.

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission_test.go (hunk 2 above)
Comment (vaikas-google) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission_test.go]: Totes agree if there are others, seems more clutter to add another package right now just for this one function however? If you feel strongly though, I can do it.

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go (hunk 1 above)
Comment (vaikas-google) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go]: talked offline, changed as per that discussion.

File: contrib/examples/walkthrough/ups-instance-default-sp.yaml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+apiVersion: servicecatalog.k8s.io/v1alpha1
```
Comment (vaikas-google) [inline contrib/examples/walkthrough/ups-instance-default-sp.yaml]: I agree that it would be a great story, I'm just worried when a user tries to create their own instance they won't have a complete example. Any case, I'd prefer to keep this here, but again if you feel strongly, I'll yank it.

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go (hunk 3 above)
Comment (vaikas-google) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go]: would you be ok with a follow on PR for this?

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission_test.go (hunk 2 above)
Comment (kibbles-n-bytes) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission_test.go]: My thoughts were that `newMockClientForTest`, `newBroker()`, and `newInstance(namespace)`  from `namespace/lifecycle/admission_test.go` would move there, as well as `newServiceClass(name, plans...)` (and the now-duplicate `newInstance(namespace)`) from this PR.

I don't have super strong feelings about it, just that it'll pretty much have to happen once we hit a third admission controller, so might as well get it out of the way.

File: contrib/examples/walkthrough/ups-instance-default-sp.yaml (hunk 4 above)
Comment (kibbles-n-bytes) [inline contrib/examples/walkthrough/ups-instance-default-sp.yaml]: I don't feel strongly. I think in the future we will want to migrate away from using user provided services as our walkthrough example anyway, as it is a bit of a special case (potentially bundled with catalog, no auth, plan concept not making sense, doesn't tell the main "consuming third party services is easy!" OSB story).

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go (hunk 3 above)
Comment (kibbles-n-bytes) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go]: That's fine. Could you make an issue for it?

Comment (vaikas-google) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go]: https://github.com/kubernetes-incubator/service-catalog/issues/1133

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go (hunk 1 above)
Comment (kibbles-n-bytes) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go]: To document for posterity: it's a bit of a strange error case, as normally an error with the service class would still be allowed to reach the API server, and the user is responsible for getting the service class working correct so that the instance will eventually succeed.

In this case, the error is actually representative of the inability to assign a default plan due to the fetching of the service class having an error. We changed the message to more clearly address that nuance.

**Comment by duglin** [comment]
What's going to happen if someone forgets to deploy the ac?

**Comment by pmorie** [comment]
Then this won't work, just like any of the other numerous features in k8s
that depend on admission controllers.

**Comment by duglin** [comment]
yea but what will the user see? It seems like the Instance should have a condition saying that it can't proceed because the plan is empty. Otherwise the user will never know what's going on and can't try to fix it.

**Comment by duglin** [comment]
From a quick look at the code it seems like we'll just call the broker with a blank planID, which will probably result in the broker generating an error - maybe. It seems that as part of this PR we should check for an empty planID before we try to provision the instance, since the spec mandates that it not be blank, and set an appropriate Condition.

File: contrib/examples/walkthrough/ups-instance-default-sp.yaml (hunk 4 above)
Comment (duglin) [inline contrib/examples/walkthrough/ups-instance-default-sp.yaml]: @vaikas-google what do you think about moving the deployment of ups to be part of the svc-cat chart so its just deployed by default?  As @kibbles-n-bytes was saying, having a walkthru (which is meant to be a learning example) not show the use of plans might not be the best example. Of course we'd need a real broker to be deployed as part of the walkthru instead.

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	// If the plan is specified, let it through and have the controller
+	// deal with finding the right plan, etc.
+	if len(instance.Spec.PlanName) > 0 {
+		return nil
+	}
+
+	sc, err := d.scLister.Get(instance.Spec.ServiceClassName)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return errors.NewInternalError(err)
```
Comment (duglin) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go]: any reason this one isn't wrapped by the "NewForbidden" logic?

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+		}
+		msg := fmt.Sprintf("ServiceClass %q does not exist, can not figure out the default Service Plan.", instance.Spec.ServiceClassName)
+		glog.V(4).Info(msg)
+		return admission.NewForbidden(a, fmt.Errorf(msg))
```
Comment (duglin) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go]: shouldn't the fmt.Errorf() just be errors.New() since by that point in the flow "msg" isn't a string needing %'s substitutions any more?  I kind of expected lint/vet to catch it.

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+	}
+	if len(sc.Plans) > 1 {
+		msg := fmt.Sprintf("ServiceClass %q has more than one plan, PlanName must be specified", instance.Spec.ServiceClassName)
+		glog.V(4).Info(msg)
+		return admission.NewForbidden(a, fmt.Errorf(msg))
```
Comment (duglin) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go]: errors.New()

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+func (d *defaultServicePlan) SetInternalServiceCatalogInformerFactory(f informers.SharedInformerFactory) {
+	scInformer := f.Servicecatalog().InternalVersion().ServiceClasses()
+	d.scLister = scInformer.Lister()
+	d.SetReadyFunc(scInformer.Informer().HasSynced)
+}
+
+func (d *defaultServicePlan) Validate() error {
+	if d.scLister == nil {
+		return fmt.Errorf("missing service class lister")
```
Comment (duglin) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go]: errors.New()

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission_test.go [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+	}
+	informerFactory.Start(wait.NeverStop)
+
+	instance := newInstance("dummy")
+	instance.Spec.ServiceClassName = "foo"
+
+	err = handler.Admit(admission.NewAttributesRecord(&instance, nil, servicecatalog.Kind("Instance").WithVersion("version"), instance.Namespace, instance.Name, servicecatalog.Resource("instances").WithVersion("version"), "", admission.Create, nil))
+	if err == nil {
+		t.Errorf("unexpected success with no ServiceClasses.List succeeding")
+	} else if !strings.Contains(err.Error(), "not yet ready to handle request") {
```
Comment (duglin) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission_test.go]: I can't express how much doing strcmp on error string like this in golang disturbs me :-(    nothing to change, just venting

**Comment by pmorie** [comment]
If a service instance reaches validation without a plan, it should be
rejected.

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go (hunk 6 above)
Comment (jpeeler) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go]: Just FYI: vet doesn't detect this because they claim it's a style preference, not one of correctness (https://github.com/golang/go/issues/17173).

Comment (duglin) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go]: that's really weird because I know I've seen some complaint from some golang checking tool complain about using things like Printf() w/o any variable args.  Maybe it was just a dream... :-)

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go (hunk 5 above)
Comment (vaikas-google) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go]: done.

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go (hunk 6 above)
Comment (vaikas-google) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go]: ack

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go (hunk 7 above)
Comment (vaikas-google) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go]: done.

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go (hunk 8 above)
Comment (vaikas-google) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go]: done.

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission_test.go (hunk 9 above)
Comment (vaikas-google) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission_test.go]: ack

**Comment by kibbles-n-bytes** [comment]
@duglin As Paul said, the apiserver validates that the plan is non-empty, so we wouldn't call the broker.

When we have more mature docs, it'd probably be helpful to mention in the section where we say that the catalog defaults to a plan if there's only one choice something like: ```if you're getting an error saying "planName is required" when instantiating a service from a class with only a single plan, make sure that the `DefaultServicePlan` admission controller is enabled.```

**Comment by duglin** [comment]
@kibbles-n-bytes - agreed, I just couldn't find that spot in the code yesterday.  Got a file/line ref?

**Comment by kibbles-n-bytes** [comment]
@duglin Lines 56-57 in `pkg/apis/servicecatalog/validation/instance.go` :slightly_smiling_face:

**Comment by vaikas-google** [comment]
I believe I've addressed all the comments, PTAL.

**Comment by duglin** [comment]
@vaikas-google thanks for the ptr

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+		return admission.NewForbidden(a, errors.New(msg))
+	}
+	if len(sc.Plans) > 1 {
+		msg := fmt.Sprintf("ServiceClass %q has more than one plan, PlanName must be specified", instance.Spec.ServiceClassName)
+		glog.V(4).Info(msg)
+		return admission.NewForbidden(a, errors.New(msg))
+	}
+
+	p := sc.Plans[0]
+	glog.V(4).Infof("Using default plan %s for Service Class %s for instance %s",
```
Comment (duglin) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go]: can you convert the %s here, and elsewhere, to %q's so things are quoted and easier to read?

**Comment by duglin** [comment]
one very minor nit and then its good to me

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go (hunk 10 above)
Comment (vaikas-google) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go]: done.

**Comment by vaikas-google** [comment]
Thanks for the reviews, PTAL.

**Comment by duglin** [comment]
mucho thanks!  LGTM

**Comment by duglin** [comment]
just waiting on CI then I'll merge

**Comment by kibbles-n-bytes** [comment]
Just FYI @duglin, @arschles , it looks like there's a backlog of work in Travis at the moment since we share our concurrent build allotment across all the incubator projects, so it will probably take quite a while before Travis actually runs.

Jenkins passed, though, so up to you whether you feel that's enough!

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission_test.go (hunk 2 above)
Comment (kibbles-n-bytes) [inline plugin/pkg/admission/serviceplan/defaultserviceplan/admission_test.go]: @arschles Yeah that's totally fine.

**Comment by arschles** [comment]
@kibbles-n-bytes let's wait for travis. @vaikas-google I will merge this when it's ready to go. If you happen to be watching it, let me know when it's ready

File: plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go (hunk 3 above)
Comment (kibbles-n-bytes) [reply plugin/pkg/admission/serviceplan/defaultserviceplan/admission.go]: :cry:
